builtin/providers/digitalocean/resource_digitalocean_firewall.go

```diff
@@ -0,0 +1,314 @@
+						},
+						"port_range": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"source_addresses": {
+							Type:     schema.TypeList,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Optional: true,
+						},
+						"source_tags": {
+							Type:     schema.TypeList,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Optional: true,
+						},
+						"source_droplet_ids": {
+							Type:     schema.TypeList,
+							Elem:     &schema.Schema{Type: schema.TypeInt},
+							Optional: true,
+						},
+						"source_load_balancer_uids": {
+							Type:     schema.TypeList,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"outbound_rules": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"protocol": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"port_range": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"destination_addresses": {
+							Type:     schema.TypeList,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Optional: true,
+						},
+						"destination_tags": {
+							Type:     schema.TypeList,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Optional: true,
+						},
+						"destination_droplet_ids": {
+							Type:     schema.TypeList,
+							Elem:     &schema.Schema{Type: schema.TypeInt},
+							Optional: true,
+						},
+						"destination_load_balancer_uids": {
+							Type:     schema.TypeList,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceDigitalOceanFirewallCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*godo.Client)
+
+	opts, err := firewallRequest(d, client)
+	if err != nil {
+		return fmt.Errorf("Error in firewall request: %s", err)
+	}
+
+	log.Printf("[DEBUG] Firewall create configuration: %#v", opts)
+
+	firewall, _, err := client.Firewalls.Create(context.Background(), opts)
+	if err != nil {
+		return fmt.Errorf("Error creating firewall: %s", err)
+	}
+
+	// Assign the firewall id
+	d.SetId(firewall.ID)
+
+	log.Printf("[INFO] Firewall ID: %s", d.Id())
+
+	return resourceDigitalOceanFirewallRead(d, meta)
+}
+
+func resourceDigitalOceanFirewallRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*godo.Client)
+
+	// Retrieve the firewall properties for updating the state
+	firewall, resp, err := client.Firewalls.Get(context.Background(), d.Id())
+	if err != nil {
+		// check if the firewall no longer exists.
+		if resp != nil && resp.StatusCode == 404 {
+			log.Printf("[WARN] DigitalOcean Firewall (%s) not found", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error retrieving firewall: %s", err)
+	}
+
+	d.Set("name", firewall.Name)
```

Comment:
Got you, Paul!  I'll add that and update the PR shortly!  Thank you for the thorough review, Paul!